packages/api-client-core/src/support.ts

```diff
@@ -495,3 +495,193 @@ export const storageAvailable = (type: "localStorage" | "sessionStorage") => {
+        }
+        fn = this.queue.shift();
+      }
+
+      this.isDequeuing = false;
+    })();
+  }
+
+  /**
+   * @param fn The function to enqueue
+   * @returns A promise that resolves when the function is dequeued and finished executing
+   */
+  enqueue<T>(fn: () => T | PromiseLike<T>): Promise<T> {
+    return new Promise((resolve, reject) => {
+      this.queue.push(async () => {
+        try {
+          resolve(await fn());
+        } catch (error) {
+          reject(error);
+        }
+      });
+
+      this.dequeue();
+    });
+  }
+}
+
+export interface TokenBucketOptions {
+  /**
+   * The capacity of the bucket
+   *
+   * @default 100
+   */
+  bucketSize: number;
+
+  /**
+   * How many tokens to add to the bucket per {@linkcode intervalMS}
+   *
+   * @default 1
+   */
+  tokensPerInterval: number;
+
+  /**
+   * The interval in ms at which tokens are added to the bucket
+   *
+   * @default 10
+   */
+  intervalMS: number;
+}
+
+/**
+ * A token bucket based on [`stopcock`](https://github.com/lpinca/stopcock)
+ *
+ * @see https://github.com/lpinca/stopcock/blob/26aa34c38bf7892e343f6f77a3dfb4c792e54d2f/index.js
+ * @license
+ * Copyright (c) 2017 Luigi Pinca
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+export class TokenBucket {
+  /**
+   * The options used to construct the bucket.
+   *
+   * @readonly This property is frozen and cannot be changed.
+   */
+  options: TokenBucketOptions;
+
+  /**
+   * The number of tokens currently in the bucket.
+   */
+  private tokens: number;
+
+  /**
+   * A `Date.now()` of the last time the bucket was refilled.
+   */
+  private lastRefill: number;
+
+  /**
+   * A queue of attempts to consume a token.
+   */
+  private queue = new PQueue();
+
+  constructor(options?: Partial<TokenBucketOptions>) {
+    this.options = Object.freeze({
+      bucketSize: 100,
+      tokensPerInterval: 1,
+      intervalMS: 10,
+      ...options,
+    });
```

Comment:
I ended up with these defaults based on rewinds 10:45 slam on Wednesday. It looks like they sent ~1150 GraphQL operations in a second across 7 sandboxes, about ~160 operations per second per sandbox.

https://app.axiom.co/gadget-vbq7/datasets/core-production?qid=qDAox1Xbh0I-s09ttw
https://ui.honeycomb.io/gadget/datasets/gadget-production/result/pULNQJe3DGB?hideMarkers

To avoid that many operations per second, but still enough to not throttle them completely, this sets the rate limit to 100 operations per second, with a max burst of 100. This can be overridden by the area config.